ocdata/core/multi_adsorbate_slab_config.py

```diff
@@ -0,0 +1,269 @@
+import copy
+import logging
+import random
+import warnings
+from itertools import product
+from typing import List
+
+import ase
+import numpy as np
+import scipy
+from ase import Atoms
+from ase.data import atomic_numbers, covalent_radii
+from ase.geometry import wrap_positions
+from pymatgen.analysis.adsorption import AdsorbateSiteFinder
+from pymatgen.io.ase import AseAtomsAdaptor
+from scipy.optimize import fsolve
+
+from ocdata.core import Adsorbate, AdsorbateSlabConfig, Slab
+from ocdata.core.adsorbate import randomly_rotate_adsorbate
+
+
+class MultipleAdsorbateSlabConfig(AdsorbateSlabConfig):
+    """
+    Class to represent a slab with multiple adsorbates on it. This class only
+    returns a fixed combination of adsorbates placed on the surface. Unlike
+    AdsorbateSlabConfig which enumerates all possible adsorbate placements, this
+    problem gets combinatorially large.
+    """
+
+    def __init__(
+        self,
+        slab: Slab,
+        adsorbates: List[Adsorbate],
+        num_sites: int = 100,
+        num_configurations: int = 1,
+        interstitial_gap: float = 0.1,
+        mode: str = "random_site_heuristic_placement",
+    ):
+        assert mode in ["random", "heuristic", "random_site_heuristic_placement"]
+        assert interstitial_gap < 5 and interstitial_gap >= 0
+
+        self.slab = slab
+        self.adsorbates = adsorbates
+        self.num_sites = num_sites
+        self.interstitial_gap = interstitial_gap
+        self.mode = mode
+
+        self.sites = self.get_binding_sites(num_sites)
+        self.atoms_list, self.metadata_list = self.place_adsorbates_on_sites(
+            self.sites,
+            num_configurations,
+            interstitial_gap,
+        )
+
+    def place_adsorbate_on_site(
```

Comment:
could we require the adsorbate be passed to the function in the base class? This feels redundant?

Reply:
I'm fine with that, in the base class we could then feed in `self.adsorbate` to this argument. @abhshkdz thoughts against this?